### PR TITLE
added support for multiple diffs (up to 8)

### PR DESCRIPTION
**Pull request by hcgraf**
OK, that was quick :) 
I just couldn't resist to hack with this a bit, so here is my patch for my improvement proposal #26 

But it was done in a few hours, with only very basic vimscript knowledge, and without in-depth testing.
So maybe this PR is not really a PR but rather a TR (Try-Request).

I tried to be minimally-invasive, to change as little as possible.
Notably, the original UI was kept, so :Linediff will automatically start the diffing after 2 blocks have been selected.

Here is the rest of my commit message, it explains how to use the modification, and some other details:

- use LinediffAdd for each Block, LinediffLast for the last one
  OR use LinediffShow AFTER the last block
- max number of diff blocks = 8 is hardcoded, because it's the max number
  of diff buffers in the  newest vim, version 8
- older versions support only 4 diff buffers, no check is done for that
  they just throw an error if you try to diff more than 4 blocks
- LinediffMerge *should* work, not tested yet at all!!
- in general not very well tested yet, use with caution!

I would be very happy to receive feedback from anybody!

### CODE REVIEW

**Comment by AndrewRadev** [comment]
At a first glance, it looks really good! I still have to test it out carefully and maybe think how the documentation will change, but we'll see. If you can keep using this branch and make modifications as you find issues, this would be a good exercise for it. In the meantime, I'll see what I can do about adding some automated tests in master.

**Comment by hcgraf** [comment]
Thank you for your feedback. I checked LinediffMerge & LinediffPick now, and made them work in the last commit.
I'm also planning to add support for git's diff3 conflictmarker-style, but I won't have time for that in the next couple of days.

In general, I'm using my branch without issues so far, but I have to admit that I'm not using it too extensively at the moment. A few automated testcases would definitely be nice!

**Comment by AndrewRadev** [comment]
I've added some automated tests. If you could merge master in it, Travis should let us know if your branch continues to work for the existing use cases (provided I've written enough tests to cover them, of course, but even if not, at least it'll cover some of the basics).

**Comment by hcgraf** [comment]
OK it fails :)
However, the "no registered server" error doesn't really seem related to my changes.
Can you comment on how to debug this?

**Comment by AndrewRadev** [comment]
Hm, that's very strange. Locally, I checked out this branch, and the build seems to run, not sure what's happening on the build server. Might be an issue with the Vim version. I'll try to debug it tomorrow, or over the weekend. I compile Vim from source, so I can check out that particular version and see what's happening.

**Comment by hcgraf** [comment]
This is now a nice new feature, LinediffMerge produces 3 diff-splits (only) when the common ancestor marker `|||||||` is present in the git merge conflict. 
Set "merge.conflictStyle" configuration variable to "diff3" to test it.

Editing and LinediffPick will of course work in all of the three splits.

The tests are still failing, though :(

**Comment by AndrewRadev** [comment]
I think I figured it out, sort of. This is a problem I used to have a long time ago, when I was trying to implement auto-LinediffReset upon closing any one of the special buffers. Turns out, Vim isn't happy about trying to close an already-closed buffer. That's why I had this code:

``` vim
" The closing logic is a bit roundabout, since changing a buffer in a
" BufUnload autocommand doesn't seem to work.
"
" The process is: if a window is entered after the other differ was destroyed,
" destroy this one as well and close the window.

  " .. .snip ...

  call s:differ_one.CreateDiffBuffer(g:linediff_first_buffer_command)
  autocmd BufUnload <buffer> silent call s:differ_one.Reset()
  autocmd WinEnter <buffer> if s:differ_two.IsBlank() | silent call s:differ_one.CloseAndReset(0) | endif

  call s:differ_two.CreateDiffBuffer(g:linediff_second_buffer_command)
  autocmd BufUnload <buffer> silent call s:differ_two.Reset()
  autocmd WinEnter <buffer> if s:differ_one.IsBlank() | silent call s:differ_two.CloseAndReset(0) | endif
```

So, if one buffer is unloaded, its data structure is reset, but the other ones get reset when the cursor ends up in them, so it happens after the autocommand is done. To reproduce, it should be enough to `:tabclose` on any one of the proxy buffers, it gets pretty confused and says it can't close the last window. There's also a situation where it segfaults, but I can't seem to figure out how to reproduce it (this is how the tests fail)

This doesn't seem to be a problem in a recent enough Vim version, but Travis uses some sensible Debian, so I'd rather not break the plugin for it -- it's likely that a lot of people use that version.

One solution I can think of is to extract some sort of diff controller, or something -- a piece of global state that can be consulted. Something like: when a buffer is closed, that controller gets a signal to propagate the effect (a flag is set, basically). Every diff buffer checks on `WinEnter` whether it should be closing itself. So, basically, the same logic, just generalized for a list of buffers rather than specifically for two. (Note that `b:differ` no longer seems to exist in the `BufUnload` event, so that might also be a problem that needs a workaround).

I can't really say how this would work without trying it out. I've been poking around trying to figure out a workaround for a while, but nothing simple seems to work. Reimplementing the old hack with some controller object should hopefully fix the issue, but there might be more complications. Since I can compile 7.4.52 locally, I'll see if I can implement it and push it to your branch.

**Comment by hcgraf** [comment]
Ok, thanks for your effort!
It's really strange, because it is [checked](https://github.com/hcgraf/linediff.vim/blob/multiple-diffs/autoload/linediff/differ.vim#L199) if the buffer exists, before it is deleted.
I tried to avoid autocmd-loops by removing them [here](https://github.com/hcgraf/linediff.vim/blob/multiple-diffs/autoload/linediff.vim#L38).

After reading a bit documentation, I saw that `BufUnload` is called **before** unloading the buffer, so the problem might rather be that the same buffer is destroyed again in that autocmd. That might be easily catchable [here](https://github.com/hcgraf/linediff.vim/blob/multiple-diffs/autoload/linediff.vim#L41).
I'm not able to try that right now because of time and not having a vim version at hand which let's me reproduce the issue…

**Comment by AndrewRadev** [comment]
Yeah, I really think it's a bug in Vim, and I think it's been fixed already (this doesn't happen in my current Vim version, or in yours). I think you've done a good job of setting up the logic, but, in the end, it's just a matter of squeezing in a hack to work around a system bug :).

I'll try to do it in the following days, but no guarantees when I'll manage, I'm afraid.

**Comment by AndrewRadev** [comment]
Okay, so, I've pushed a few commits to your branch which should fix the tests. I've basically reimplemented the old `WinEnter` logic, and boy, was it annoying. It's also super hacky and it doesn't work for all scenarios, but I've got another workaround for it.

For now, what I still need to do is:
- Write some documentation on the closing hacks, because I *really* don't want to figure the scenarios out again next time.
- Write some tests on the new commands, and the new merge behaviour with the threeway merge.

If you can test this on your own Vim version and let me know if it doesn't segfault or something, that would be great. I'll switch to Vim 8 soon as well, but I'd like to write the tests first, so I know the new ones continue to work.

**Comment by hcgraf** [comment]
cool, looks like you thought about a lot of corner cases!
I will most probably get some opportunities tomorrow to test it in a real workflow. I'll report back!

Btw, I was using/testing my own version quite a bit over the last weeks, and it worked fine but I got some error messages when closing modified diff-tabs.
For me, that's OK as long as I know what's going on behind the scenes (I just ignored the errors), but it's of course not ready for a release. I've never seen a segfault though.

**Comment by hcgraf** [comment]
I didn't have too many real-world testcases these two days, but some. No problems so far!

I will definitely continue to use this branch, so I'll report any possible problems here.

Thank you very much for your effort!

File: autoload/linediff/differ.vim [hunk 1]
```diff
@@ -12,8 +12,10 @@ function! linediff#differ#New(sign_name, sign_number)
         \ 'sign_number':        a:sign_number,
         \ 'sign_text':          a:sign_number.'-',
         \ 'is_blank':           1,
-        \ 'other_differ':       {},
+        \ 'other_differs':      {},
```
Comment (hcgraf) [inline autoload/linediff/differ.vim]: Very well. I did all this basically without vimscript knowledge, only general programming knowledge. So it was mostly a mix'n'match approach ;)

**Comment by hcgraf** [comment]
Wow, thanks for the continued effort!
I glanced over most of your changes when you pushed them, and everything seems fine! But I didn't have enough time to really analyze it & do some tests.

I'm currently using it at work on a regular basis, but usually in a similar flow - so it's not likely to cover strange corner cases. Anyway, no issues so far! (Btw this is GVim in Windows and regular Vim in Linux - so at least some cases covered).

I think I understand the logic of the closing-logic, the documentation seems as clear as it can get for this one. Although I recently had one use-case where I had wished for another behavior:
- I was diffing three regions
- I wanted to close one of them and execute `:diffupdate` to only compare two of them
- That was of course not possible, as closing one closes all
- But that's a different story (maybe a next feature-request), and shouldn't be related to this PR

As for merging: I think it's highly unlikely that this PR will destroy anything for the regular user (used to diffing two areas and continuing to use it like this), so it might be helpful to get some more field-tests for the N-way diffing…
So I would probably vote for merging.

**Comment by AndrewRadev** [comment]
Yes, I can agree that closing one buffer to leave two makes sense. Maybe it's time to reconsider the logic there. Originally, the idea was that there's no point in leaving just one diff buffer. Now that it's possible to add a bunch of them, it doesn't hold up.

If you happen to have some time, I would love to get a PR for this logic, not sure how hard it would be. Something like it only resets once there's only one differ left. If not, an issue with a feature request to remind me would be fine, and I'd get to working on it in my own time.

Either way, time to merge this and see what's next :). Thank you very much for your work!
